src/devices/src/virtio/balloon/device.rs

```diff
@@ -0,0 +1,920 @@
+    pub(crate) queues: Vec<Queue>,
+    pub(crate) interrupt_status: Arc<AtomicUsize>,
+    interrupt_evt: EventFd,
+    pub(crate) queue_evts: [EventFd; NUM_QUEUES],
+    pub(crate) device_state: DeviceState,
+
+    // Implementation specific fields.
+    pub(crate) stats_polling_interval_s: u16,
+    pub(crate) stats_timer: TimerFd,
+    // The index of the previous stats descriptor is saved because
+    // it is acknowledged after the stats queue is processed.
+    pub(crate) stats_desc_index: Option<u16>,
+    pub(crate) latest_stats: BalloonStats,
+}
+
+impl Balloon {
+    pub fn new(
+        num_pages: u32,
+        must_tell_host: bool,
+        deflate_on_oom: bool,
+        stats_polling_interval_s: u16,
+    ) -> Result<Balloon, Error> {
+        let mut avail_features = 1u64 << VIRTIO_F_VERSION_1;
+
+        if must_tell_host {
+            avail_features |= 1u64 << VIRTIO_BALLOON_F_MUST_TELL_HOST;
+        };
+
+        if deflate_on_oom {
+            avail_features |= 1u64 << VIRTIO_BALLOON_F_DEFLATE_ON_OOM;
+        };
+
+        if stats_polling_interval_s > 0 {
+            avail_features |= 1u64 << VIRTIO_BALLOON_F_STATS_VQ;
+        }
+
+        let queue_evts = [
+            EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+            EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+            EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+        ];
+
+        let queues = QUEUE_SIZES.iter().map(|&s| Queue::new(s)).collect();
+
+        let stats_timer = TimerFd::new_custom(ClockId::Monotonic, true, true).map_err(Error::IO)?;
+
+        Ok(Balloon {
+            avail_features,
+            acked_features: 0u64,
+            config_space: ConfigSpace {
+                num_pages,
+                actual_pages: 0,
+            },
+            interrupt_status: Arc::new(AtomicUsize::new(0)),
+            interrupt_evt: EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+            queue_evts,
+            queues,
+            device_state: DeviceState::Inactive,
+            activate_evt: EventFd::new(libc::EFD_NONBLOCK).map_err(Error::EventFd)?,
+            stats_polling_interval_s,
+            stats_timer,
+            stats_desc_index: None,
+            latest_stats: BalloonStats::default(),
+        })
+    }
+
+    pub(crate) fn process_inflate_queue_event(&mut self) {
+        if let Err(e) = self.queue_evts[INFLATE_INDEX].read() {
+            error!("Failed to get queue event: {:?}", e);
+            METRICS.balloon.event_fails.inc();
+        } else {
+            self.process_inflate()
+                .unwrap_or_else(report_balloon_event_fail);
+        }
+    }
+
+    pub(crate) fn process_deflate_queue_event(&mut self) {
+        if let Err(e) = self.queue_evts[DEFLATE_INDEX].read() {
+            error!("Failed to get queue event: {:?}", e);
+            METRICS.balloon.event_fails.inc();
+        } else if self.process_deflate_queue() {
```

Comment:
Don't we need `report_balloon_event_fail` in case `process_deflate_queue()` fails?

Reply:
We do not as all operations in `process_deflate_queue()` are safe, the only things that can fail when dealing with the deflate queue are failure to read the `EventFd` and failure to signal the used queue.

The boolean return value corresponds to whether or not an interrupt is needed after processing the deflate queue, not whether or not the processing has succeeded.